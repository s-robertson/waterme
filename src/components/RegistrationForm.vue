<template>
  <div class="c-registration-form">
    <ErrorList :errors="registrationErrors" />
    <div v-if="errorMessage">
      {{ errorMessage }}
    </div>
    <form @submit.prevent="handleSubmit">
      <div class="c-form-item">
        <label for="email">
          Email address:
          <input v-model="email" id="email" type="text" />
        </label>
      </div>
      <div class="c-form-item">
        <label for="password">
          Password:
          <input v-model="password" id="password" type="password" />
        </label>
      </div>
      <button type="submit">Register</button>
    </form>
  </div>
</template>

<script>
import { mapActions, mapState } from "vuex";
import ErrorList from "@/components/ErrorList";

export default {
  name: "RegistrationForm",
  components: { ErrorList },
  data() {
    return {
      email: "",
      password: ""
    };
  },
  computed: {
    ...mapState(["registrationErrors"])
  },
  methods: {
    ...mapActions(["register"]),
    async handleSubmit() {
      const result = await this.register({
        email: this.email,
        password: this.password
      });
      if (result) {
        this.$router.push("/");
      }
    }
  }
};
</script>
