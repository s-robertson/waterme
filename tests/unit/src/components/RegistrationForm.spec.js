import { createLocalVue, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";
import RegistrationForm from "@/components/RegistrationForm";
import { createStore } from "../../helpers";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("RegistrationForm component", () => {
  const defaultStoreOptions = {
    state: {
      registrationErrors: []
    },
    actions: {
      register: jest.fn()
    }
  };

  it("Should render correctly", () => {
    const wrapper = shallowMount(RegistrationForm, {
      localVue,
      store: createStore(defaultStoreOptions)
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should display any registration errors", () => {
    const wrapper = shallowMount(RegistrationForm, {
      localVue,
      store: createStore(defaultStoreOptions, {
        state: { registrationErrors: ["Invalid email"] }
      })
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should register a new user", async () => {
    const expectedEmail = "test@user.com";
    const expectedPassword = "myPassword123";
    defaultStoreOptions.actions.register.mockImplementation(async () => {
      return true;
    });
    const $router = {
      push: jest.fn()
    };

    const wrapper = shallowMount(RegistrationForm, {
      localVue,
      store: createStore(defaultStoreOptions),
      mocks: {
        $router
      }
    });

    wrapper.find("input[type=text]").setValue(expectedEmail);
    wrapper.find("input[type=password]").setValue(expectedPassword);
    wrapper.find("form").trigger("submit");

    await wrapper.vm.$nextTick();

    expect($router.push).toHaveBeenCalledWith("/");
    expect(defaultStoreOptions.actions.register).toHaveBeenCalledWith(
      expect.anything(),
      {
        email: expectedEmail,
        password: expectedPassword
      }
    );
  });
});
