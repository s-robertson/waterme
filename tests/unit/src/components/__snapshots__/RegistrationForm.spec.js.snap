// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm component Should display any registration errors 1`] = `
<div class="c-registration-form">
  <errorlist-stub errors="Invalid email"></errorlist-stub>
  <!---->
  <form>
    <div class="c-form-item"><label for="email">
        Email address:
        <input id="email" type="text"></label></div>
    <div class="c-form-item"><label for="password">
        Password:
        <input id="password" type="password"></label></div> <button type="submit">Register</button>
  </form>
</div>
`;

exports[`RegistrationForm component Should render correctly 1`] = `
<div class="c-registration-form">
  <errorlist-stub errors=""></errorlist-stub>
  <!---->
  <form>
    <div class="c-form-item"><label for="email">
        Email address:
        <input id="email" type="text"></label></div>
    <div class="c-form-item"><label for="password">
        Password:
        <input id="password" type="password"></label></div> <button type="submit">Register</button>
  </form>
</div>
`;
